gpflow/optimizers/scipy.py

```diff
@@ -34,44 +34,65 @@ def minimize(self,
         if not callable(closure):
             raise TypeError('Callable object expected.')  # pragma: no cover
         initial_params = self.initial_parameters(variables)
-        func = self.eval_func(closure, variables, step_callback)
+        func = self.eval_func(closure, variables)
+        if step_callback is not None:
+            if 'callback' in scipy_kwargs:
+                raise ValueError("Callback passed both via `step_callback` and `callback`")
+
+            callback = self.callback_func(closure, variables, step_callback)
+            scipy_kwargs.update(dict(callback=callback))
         return scipy.optimize.minimize(func, initial_params, jac=True, **scipy_kwargs)
 
     @classmethod
     def initial_parameters(cls, variables):
         return cls.pack_tensors(variables)
 
     @classmethod
-    def eval_func(cls, closure: LossClosure, variables: Variables, step_callback: Optional[StepCallback] = None):
-        step = 0  # type: int
-
+    def eval_func(cls, closure: LossClosure, variables: Variables):
         def _eval(x):
-            nonlocal step
-            cls.unpack_tensors(variables, x)
+            values = cls.unpack_tensors(variables, x)
+            cls.assign_tensors(variables, values)
```

Comment:
This will lead to a huge slowdown. The idea behind step_callback: the user decides when to do assigning of values to the models.

Reply:
This is inside the eval_func, not the callback_func.

Reply:
it's doing the same assign as before - just more explicitly!